ReSharper.FSharp/src/FSharp.Psi.Services/src/Util/SignatureFileUtil.fs

```diff
@@ -0,0 +1,172 @@
+        match nmd.Parent with
+        | :? IFSharpSigFile -> true
+        | _ -> false) allDecls
+
+    let elementFactory = sigDecl.CreateElementFactory()
+    let binding = letBindings.Bindings.First()
+    let refPat = binding.HeadPattern.As<IReferencePat>()
+    let name = refPat.ReferenceName.Identifier.Name
+    let symbolUse = refPat.GetFcsSymbolUse()
+    if isNull symbolUse then None else
+
+    let mfv = symbolUse.Symbol :?> FSharpMemberOrFunctionOrValue
+    let types = FcsTypeUtil.getFunctionTypeArgs true mfv.FullType
+    let parameters = binding.ParametersDeclarations
+
+    let rec getTypeName (t:FSharpType) : string =
+        if t.HasTypeDefinition then
+            if Seq.isEmpty t.GenericArguments then
+                t.TypeDefinition.DisplayName
+            else
+                let isPostFix =
+                    set [| "list"; "option"; "array" |]
+                    |> Set.contains t.TypeDefinition.DisplayName
+                    
+                if isPostFix then
+                    let ga = t.GenericArguments[0].GenericParameter
+                    let tick = if ga.IsSolveAtCompileTime then "^" else "'"
+                    $"{tick}{ga.DisplayName} {t.TypeDefinition.DisplayName}"
+                else
+                    let args = Seq.map getTypeName t.GenericArguments |> String.concat ","
+                    $"{t.TypeDefinition.DisplayName}<{args}>"
+        elif t.IsGenericParameter then
+            let tick = if t.GenericParameter.IsSolveAtCompileTime then "^" else "'"
+            $"{tick}{t.GenericParameter.DisplayName}"
+        elif t.IsFunctionType then
+            let rec visit (t: FSharpType) : string seq =
+                if  t.IsFunctionType then
+                    Seq.collect visit t.GenericArguments
+                else
+                    Seq.singleton (getTypeName t)
+            
+            visit t
+            |> String.concat " -> "
+            |> sprintf "(%s)"
+        elif t.IsTupleType then
+            t.GenericArguments
+            |> Seq.map getTypeName
+            |> String.concat " * "
+        else
+            "???"
+
+    let rec visitAccessPath (t: FSharpType) : string list =
+        [
+            if t.HasTypeDefinition then
+                yield t.TypeDefinition.AccessPath
+            if not t.IsGenericParameter && not (Seq.isEmpty t.GenericArguments) then
+                yield! (Seq.collect visitAccessPath t.GenericArguments)
+        ]
+
+    let alwaysOpen =
+        let existingOpenStatements =
+            sigDecl.Members
+            |> Seq.choose (function | :? IOpenStatement as os -> Some os | _ -> None)
+            |> Seq.map (fun openStatement -> openStatement.ReferenceName.QualifiedName)
+
+        set
+            [|
+                yield "Microsoft.FSharp.Core"
+                yield "Microsoft.FSharp.Collections"
+                yield sigDecl.CompiledName
+                yield! existingOpenStatements
+            |]
```

Comment:
Let's try to use `OpensUtil.OpenedModulesProvider.OpenedModuleScopes` instead? It should properly deal with all implicitly opened modules, take into account the current one declaration and opened `open` statements.